tests/e2e/p/permissionless_layer_one.go

```diff
@@ -206,6 +247,198 @@ var _ = e2e.DescribePChain("[Permissionless L1]", func() {
+					registerSubnetValidatorMessage.Bytes(),
+				)).Bytes(),
+			))
+
+			tc.By("sending the request to sign the warp message", func() {
+				registerSubnetValidatorRequest, err := wrapWarpSignatureRequest(
+					unsignedRegisterSubnetValidator,
+					nil,
+				)
+				require.NoError(err)
+
+				require.True(genesisPeer.Send(tc.DefaultContext(), registerSubnetValidatorRequest))
+			})
+
+			tc.By("getting the signature response")
+			registerSubnetValidatorSignature, ok, err := findMessage(genesisPeerMessages, unwrapWarpSignature)
+			require.NoError(err)
+			require.True(ok)
+
+			tc.By("creating the signed warp message to register the validator")
+			signers := set.NewBits()
+			signers.Add(0) // [signers] has weight from the genesis peer
+
+			var sigBytes [bls.SignatureLen]byte
+			copy(sigBytes[:], bls.SignatureToBytes(registerSubnetValidatorSignature))
+			registerSubnetValidator, err := warp.NewMessage(
+				unsignedRegisterSubnetValidator,
+				&warp.BitSetSignature{
+					Signers:   signers.Bytes(),
+					Signature: sigBytes,
+				},
+			)
+			require.NoError(err)
+
+			tc.By("issuing a RegisterSubnetValidatorTx", func() {
+				_, err := pWallet.IssueRegisterSubnetValidatorTx(
+					registerBalance,
+					registerNodePoP.ProofOfPossession,
+					registerSubnetValidator.Bytes(),
+				)
+				require.NoError(err)
+			})
+		})
+
+		tc.By("verifying the validator was registered", func() {
+			tc.By("verifying the validator set was updated", func() {
+				verifyValidatorSet(map[ids.NodeID]*snowvalidators.GetValidatorOutput{
+					subnetGenesisNode.NodeID: {
+						NodeID:    subnetGenesisNode.NodeID,
+						PublicKey: genesisNodePK,
+						Weight:    genesisWeight,
+					},
+					ids.EmptyNodeID: { // The validator is not active
+						NodeID: ids.EmptyNodeID,
+						Weight: registerWeight,
+					},
+				})
+			})
+		})
+
 		_ = e2e.CheckBootstrapIsPossible(tc, env.GetNetwork())
 	})
 })
+
+func wrapWarpSignatureRequest(
+	msg *warp.UnsignedMessage,
+	justification []byte,
+) (p2pmessage.OutboundMessage, error) {
+	p2pMessageFactory, err := p2pmessage.NewCreator(
+		logging.NoLog{},
+		prometheus.NewRegistry(),
+		constants.DefaultNetworkCompressionType,
+		10*time.Second,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	request := sdk.SignatureRequest{
+		Message:       msg.Bytes(),
+		Justification: justification,
+	}
+	requestBytes, err := proto.Marshal(&request)
+	if err != nil {
+		return nil, err
+	}
+
+	return p2pMessageFactory.AppRequest(
+		msg.SourceChainID,
+		0,
+		time.Hour,
+		p2psdk.PrefixMessage(
+			p2psdk.ProtocolPrefix(p2psdk.SignatureRequestHandlerID),
+			requestBytes,
+		),
+	)
+}
+
+func findMessage[T any](
+	q buffer.BlockingDeque[p2pmessage.InboundMessage],
+	parser func(p2pmessage.InboundMessage) (T, bool, error),
+) (T, bool, error) {
+	var messagesToReprocess []p2pmessage.InboundMessage
+	defer func() {
+		slices.Reverse(messagesToReprocess)
+		for _, msg := range messagesToReprocess {
+			q.PushLeft(msg)
```

Comment:
nit: could we go in order then PushRight?

Reply:
No, that could end up re-ordering messages.

For example, say we receive messages `1`, `2`, `3`, `4`.

If try to remove `3`, we want to end with `1`, `2`, `4`.

We do this by looking at `1`, then `2`, then `3`.

If we returned them in order to the other side of the queue after this, we would end up with `4`, `1`, `2`.